mdapi/MetricsDiscoveryHelper.cpp

```diff
@@ -0,0 +1,1142 @@
+    sizeof(void*) == 8 ?
+    L"igdmd64.dll" :
+    L"igdmd32.dll";
+
+#include "DriverStorePath.h"
+
+#define OpenLibrary(_filename)              (void*)LoadDynamicLibrary( _filename )
+#define GetFunctionAddress(_handle, _name)  GetProcAddress((HMODULE)_handle, _name)
+
+#elif defined(__linux__) || defined(__APPLE__)
+static const char* cMDLibFileName = "libmd.so";
+
+#include <dlfcn.h>
+#include <stdarg.h>
+#include <string.h>
+
+#define OpenLibrary(_filename)              dlopen( _filename, RTLD_LAZY | RTLD_LOCAL )
+#define GetFunctionAddress(_handle, _name)  dlsym(_handle, _name)
+#define GetLastError()                      dlerror()
+#define OutputDebugString(_buf)             fprintf(stderr, "%s\n", _buf);
+
+#endif
+
+namespace MetricsDiscovery
+{
+
+#ifdef MD_DEBUG
+/************************************************************************/
+/* DebugPrint                                                           */
+/************************************************************************/
+static void DebugPrint(const char* formatString, ...)
+{
+    const int BUFF_SIZE = 1024;
+    const int FORMAT_SIZE = 512;
+
+    va_list ap;
+    char buf[BUFF_SIZE];
+    char format[FORMAT_SIZE];
+
+    CLI_SPRINTF(format, FORMAT_SIZE, "MDAPI Helper: %s", formatString);
+    va_start(ap, formatString);
+    CLI_VSPRINTF(buf, BUFF_SIZE, format, ap);
+    OutputDebugString(buf);
+    va_end(ap);
+}
+#else
+#define DebugPrint(...)
+#endif // MD_DEBUG
+
+/************************************************************************/
+/* MDHelper constructor                                                 */
+/************************************************************************/
+MDHelper::MDHelper() :
+    m_Initialized(false),
+    m_APIMask( API_TYPE_OCL | API_TYPE_OGL | API_TYPE_DX12 ),
+    m_CategoryMask( GPU_RENDER | GPU_COMPUTE | GPU_MEDIA | GPU_GENERIC ),
+    m_MetricsDevice( NULL ),
+    m_MetricSet( NULL )
+{
+}
+
+/************************************************************************/
+/* ~MDHelper                                                            */
+/************************************************************************/
+MDHelper::~MDHelper()
+{
+    DebugPrint("MDHelper Destructor");
+
+    if(CloseMetricsDevice != NULL && m_MetricsDevice)
+    {
+        CloseMetricsDevice( m_MetricsDevice );
+        m_MetricsDevice = NULL;
+    }
+
+    m_Initialized = false;
+}
+
+/************************************************************************/
+/* InitMetricsDiscovery                                                 */
+/************************************************************************/
+bool MDHelper::InitMetricsDiscovery(
+    const std::string& metricSetSymbolName,
+    const std::string& metricsFileName )
+{
+    m_Initialized = false;
+
+    CloseMetricsDevice = NULL;
+    OpenMetricsDevice = NULL;
+    OpenMetricsDeviceFromFile = NULL;
+
+    TCompletionCode res = CC_OK;
+
+    if (m_APIMask & API_TYPE_IOSTREAM )
```

Comment:
Better check whether it's not different from OCL, because IOSTREAM isn't the only TYPE that shouldn't be used here.